#19: reducing the number of Rect allocations during list scrolling.

denisk20 opened the pull request:
This makes scrolling the demo visually smoother even on emulator.

jacobtabak commented:
Thanks so much! Was out of town at Droidcon this week, we'll review and merge next week!

denisk20 commented:
Sure, let me know if anything needs to be improved.

denisk20 commented on library/src/main/java/com/timehop/stickyheadersrecyclerview/StickyRecyclerHeadersDecoration.java:
```diff
@@ -101,9 +103,20 @@ public void onDrawOver(Canvas canvas, RecyclerView parent, RecyclerView.State st
+      }
+    }
+    //remove unused rects, if any.
+    if (childCount < mHeaderRects.size()) {
```
Really a micro-optimization here, it will work without this `if`. On the other hand, this code is executed very frequently during scrolling, so I though any optimization could matter. Let me know if you think I need to simplify this.

jacobtabak replied:
I don't really think this will be too helpful, probably not worth it for the added complexity.  If the rects were allocated once they most likely will be needed again if the user scrolls up to a point where the item density is higher.

denisk20 replied:
How about the situation when objects are removed from the list dynamically? In this case we might have inconsistent (yet probably harmless) state of `mHeaderRects` containing rects that correspond to non-existing positions in the list. This is a corner case though... It used to be solved automatically since `mHeaderRects` was cleared on every scroll move.

jacobtabak replied:
I apologize for the slow response, and also if we're not entirely on the same page. You've put a lot more thought into this than I have, so please take my comments with a grain of salt and correct me if I'm wrong.

If items are removed from the list dynamically, I'm not sure this code would help. For example if you have 4 items in the adapter, and 3 of those are on screen (e.g. `getChildCount()`), and then you remove the 3rd one. `getChildCount()` will not change - there will still be 3 views attached to the RecyclerView, but the state of `mHeaderRects` will still be stale.

But I think in the big picture it doesn't matter if the values in `mHeaderRects` are stale or the incorrect size - the correct dimensions are calculated each time the view is drawn.  

As far as I can tell, this optimization really only saves memory at the cost of a few extra allocations.

jacobtabak replied:
Curious to hear your thoughts - and let me know if I'm missing the boat. I'd love to merge this and make a release today or tomorrow!

denisk20 replied:
You're right - I forgot that there are as many `mHeaderRects` items as RecyclerView view count, not as many as adapter items count. I'll submit the change shortly. Thanks for pointing that out!

jacobtabak commented on library/src/main/java/com/timehop/stickyheadersrecyclerview/HeaderPositionCalculator.java:
```diff
@@ -19,6 +19,9 @@
   private final DimensionCalculator mDimensionCalculator;
+  
+  private final Rect bufferRect1 = new Rect();
+  private final Rect bufferRect2 = new Rect();
```
Can we use more descriptive variable names? Also we should probably use the `m` prefix (although I now hate it) to be consistent

denisk20 replied:
Does something like `mTempRect2` sound better? I can't think of any much better name...

jacobtabak replied:
Previously they were local variables and called `headerMargins` and `nextHeaderMargins`, maybe we can just call them `mHeaderRect` and `mNextHeaderRect` or something

denisk20 replied:
My intent was to leave these variable context-less, explicitly showing that they are just buffers which are useful only in context of particular method. The names that you proposed will definitely work, but if, for example, we decide to calculate footer rect we'll have to write something `initFooterRect(mHeaderRect, ...); float footerX = mHeaderRect.x;` which will be more confusing than `initFooterRect(mTempRect, ...); float footerX = mTempRect.x;`. There is another aspect: a future reader might think that `mHeaderRect` is some useful permanent field, but it will just contain some leftovers from previous calculations. `mTmpRect` suggests the latter. 
Just my thoughts...

jacobtabak commented:
Your reasoning makes sense... maybe just a javadoc for the fields explaining what they're used for would be appropriate?

denisk20 commented:
I have changed field names to `mTempRect`, added some javadocs. I have rebased so there is still one commit. I haven't addressed your comment on removing old rects since you didn't respond to my last note; let me know if I need to change it.

jacobtabak commented:
Thanks! Merging

denisk20 commented:
@jacobtabak Do you have plans to release a new version anytime soon? I'd love to incorporate these and other changes into my project. Thanks!

jacobtabak commented:
I do, there's another pending PR that has a pretty important fix but looks like the author hasn't followed up - I'll poke him !
